Feature json expression

Deecellar opened the pull request:
Added feature following https://quire.io/w/Monofoxe/11/Add_support_for_m...?board=Monofoxe&filter=all 
Gray Foxes helped me in this one

- Added Naive implementation on tokenizer/parser
- PEMDAS is followed (even tho technically the E does not exist =D)

Deecellar committed:
Added initial parser

Deecellar committed:
Added Comments, also, fixed slight codestyle issues

Deecellar committed:
Removed Unused Usings

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupImporter.cs:
```diff
@@ -131,40 +131,12 @@ private void ImportTextures(string dirPath, string dirName, SpriteGroupData grou
 
-						// Hey, look, switch is being useful for once! :000
-
+						SpriteGroupMathParser parser = new SpriteGroupMathParser(spr);
```
Use var.

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs:
```diff
@@ -0,0 +1,228 @@
+		{
+			switch (_tokenizer.CurrentToken)
+			{
+				case Token.Number:
```
This is barely readable. Add braces to cases, or better yet - get rid of the switch.

Deecellar replied:
Decided first form, not a fan of 
```csharp
if(_tokenizer.CurrentToken == Token.Number) 
{
	//...
}
```
with a lot of conditions

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs:
```diff
@@ -0,0 +1,228 @@
+					}
+					return new NumberNode(_rawSprite.RawTexture.Width / _rawSprite.FramesH / 2);
+				default:
+					throw new Exception($"Unexpect token: {_tokenizer.CurrentToken}"); // TODO: Make a SyntaxException or something similar
```
Forgot the TODO. : - )

Deecellar replied:
True =D not anymore

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs:
```diff
@@ -0,0 +1,228 @@
+	public class SpriteGroupMathParser
+	{
+
+		public SpriteGroupMathParser(RawSprite sprite)
```
Make this class static, there is no reason for it to be instantiated.

Deecellar replied:
Done =3

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs:
```diff
@@ -0,0 +1,228 @@
+
+		public SpriteGroupMathParser(RawSprite sprite)
+		{
+			_rawSprite = sprite; // Maybe this should not be here but hey... Fox are sneaky 
```
Yeah, it most definitely should not. I propose removing _rawSprite and _yAxis and passing constants directly into the Parse method.

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs:
```diff
@@ -0,0 +1,228 @@
+	}
+	#region NodeClasses
+
+	abstract class Node
```
Why are all those in a single file?

Deecellar replied:
oh my, forgot to let refactor that when cleaning

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs:
```diff
@@ -0,0 +1,134 @@
+	{
+	  _builder = new StringBuilder(text);
+	  _builder.Replace("center", "c").Replace("top", "t").Replace("bottom", "b").Replace("left", "l").Replace("right", "r");
+	  _text = _builder.ToString();
```
The string is way too longe, make it something like
```
_builder.Replace("center", "c")
  .Replace("top", "t")
  .Replace("bottom", "b")
```

Deecellar replied:
Yuop

Martenfur commented on Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs:
```diff
@@ -0,0 +1,134 @@
+	  _text = _builder.ToString();
+	  _textReader = new StringReader(_text);
+	  _builder.Clear();
+	  NextChar();
```
Those are indented with spaces, by the way. Check the rest to have the correct indents.

Deecellar replied:
Codestyle ahhh

Deecellar committed:
Fixed to Codestyle

Deecellar committed:
Modified how the offset is read

- Fixed Codestyle
- Separated Node and Token exception on the Pipefoxe.SpriteGroup.AST namespace
- SpriteGroupMathParser is now a static class
- SpriteGroupMathParser.Parse now returns the offset directly
- SpriteGroupMathParser.Parse now Accepts the constants to be passed
- Adapted to new parse method in SpriteGroupImporter
- SpriteGroupMathTokenizer passed to Pipefoxe.SpriteGroup.AST namespace

Deecellar committed:
Sln reverted to before error

I need to stop making this class of errors

Martenfur commented:
Since there were conflicts, I've just copied the code and pushed it locally. Gonna close this PR.
